src/dex/swaap-v1/swaap-v1.ts

```diff
@@ -0,0 +1,621 @@
+          );
+        } catch (e) {
+          return 0n;
+        }
+      });
+
+      return [unitAmountOut, amountsOut];
+    } else {
+      let unitAmountIn;
+      try {
+        unitAmountIn = PoolQuotations.getAmountInGivenOutMMM(
+          state.liquidities[from],
+          state.oracles[from],
+          state.liquidities[to],
+          state.oracles[to],
+          unitVolume,
+          state.parameters,
+          currentTimestamp,
+        );
+      } catch (e) {
+        unitAmountIn = BigInt(MAX_UINT);
+      }
+
+      const amountsIn = amounts.map(amountOut => {
+        try {
+          return PoolQuotations.getAmountInGivenOutMMM(
+            state.liquidities[from],
+            state.oracles[from],
+            state.liquidities[to],
+            state.oracles[to],
+            amountOut,
+            state.parameters,
+            currentTimestamp,
+          );
+        } catch (e) {
+          return BigInt(MAX_UINT);
+        }
+      });
+      return [unitAmountIn, amountsIn];
+    }
+  }
+
+  // Returns pool prices for amounts.
+  // If limitPools is defined only pools in limitPools
+  // should be used. If limitPools is undefined then
+  // any pools can be used.
+  getPricesVolume(
+    _from: Token,
+    _to: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<SwaapV1Data>> {
+    return this.getPricesVolumeLogic(
+      _from,
+      _to,
+      amounts,
+      side,
+      blockNumber,
+      null,
+      limitPools,
+    );
+  }
+
+  async getPricesVolumeLogic(
+    _from: Token,
+    _to: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    _currentTimeStamp: bigint | null,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<SwaapV1Data>> {
+    const from = this.dexHelper.config.wrapETH(_from);
+    const to = this.dexHelper.config.wrapETH(_to);
+
+    const srcTokenAddress = from.address.toLowerCase();
+    const destTokenAddress = to.address.toLowerCase();
+
+    const topPools: Address[] = this.findPools(
+      srcTokenAddress,
+      destTokenAddress,
+    );
+
+    const allowedPools = limitPools
+      ? topPools.filter((p: Address) =>
+          limitPools.includes(`${this.dexKey}_${p.toLowerCase()}`),
+        )
+      : topPools;
+
+    if (!allowedPools.length) return null;
+
+    const unitVolume = getBigIntPow(
+      (side === SwapSide.SELL ? from : to).decimals,
+    );
+
+    if (srcTokenAddress === destTokenAddress) return null;
+    const currentTimestamp =
+      _currentTimeStamp || BigInt(Math.floor(Date.now() / 1000)); // the on-chain pricing logic is not too much sensitive to that value
+    return await Promise.all(
+      allowedPools.map(async poolAddress => {
+        let state = this.eventPools[poolAddress].getState(blockNumber);
+        if (!state) {
+          state = await this.eventPools[poolAddress].generateState(blockNumber);
```

Comment:
Should call `setState` with the generated state so that it can possibly be used again.